### PR TITLE
pr-pull: check for conflicts with long running builds

### PR DESCRIPTION
**Pull request by iMichka**
This change will prevent us having to run some long running builds
multiple times and to rely on luck to get things merged without conflicts.

The check takes less than 30 seconds on my local setup.

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:386]:
```suggestion
    return if conflicts.blank?
```

File: Library/Homebrew/utils/github.rb
Comment (carlocab) [inline Library/Homebrew/utils/github.rb:511]: I feel like there's a more intuitive way to organise how we keep track of these files, instead of an `Array` of `Hash` maps.

File: Library/Homebrew/utils/github.rb
Comment (carlocab) [inline Library/Homebrew/utils/github.rb:509]: This rest query appears twice (once here, and once in `get_pr_files`). Is there some way to recycle these query results? In particular I think you query the files in a `long build` PR at least twice via the REST API.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:389]: Could you show an example error message here?

I think this might also read better:
```suggestion
    odie <<~EOS
      You are trying to merge a pull request that conflicts with
      a long running build in
        #{conflicts * "  \n"}"
    EOS
```

**Comment by Bo98** [comment]
This feels like something that should be on the `pr-automerge` layer so it doesn't even try to merge rather than try and fail.

Even then not all long build PRs are necessary always being worked on. I feel like it should be a separate label.

**Comment by iMichka** [comment]
> This feels like something that should be on the pr-automerge layer so it doesn't even try to merge rather than try and fail.

I thought about it. But this excludes 2 cases:
- maintainers running pr-pull locally (I don't know how often that happens)
- automerge-skip label, which will not use the pr-automerge process

> Even then not all long build PRs are necessary always being worked on. I feel like it should be a separate label.

I see. Some long builds get merged quite quickly, when they are short.
We can introduce a new label for the longer ones. This should also reduce the number of API call, as less PRs need to be checked. How should we call that new label? `forbid-conflict-merge`?

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb]: Done. I adjusted it and added a comment to the code.

File: Library/Homebrew/utils/github.rb
Comment (iMichka) [reply Library/Homebrew/utils/github.rb:509]: Maybe there some ruby niceness that can improve this. My goal was to find a working solution first. If anyone has a better idea I'm open to implement it :)

File: Library/Homebrew/utils/github.rb
Comment (iMichka) [reply Library/Homebrew/utils/github.rb]: I added a `get_pr_files`, that can now be used at 3 places in the file.

What do you mean by "query the files twice"?

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:394]:
```suggestion
        #{conflicts * "  \n"}
```
I think your alignment will come out wrong for multiple PRs without the extra indent.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:391]: I didn't mean here, but this is ok too 😃

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:27]: When would you want to pass this?

File: Library/Homebrew/utils/github.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb:503]: This doesn't feel right to be in this class; it's way too specific to `brew pull`.

File: Library/Homebrew/utils/github.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb:515]: Same here, I'd suggest this logic is moved into `brew bump-formula-pr`.

File: Library/Homebrew/utils/github.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb:516]:
```suggestion
    get_pr_files(name, tap_remote_repo, pr).map(&:filename)
```

may work

File: Library/Homebrew/utils/github.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb:519]:
```suggestion
  def get_pull_request_changed_files(name, tap_remote_repo, pr)
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:386]: I'd suggest inlining all this logic, removing the methods (except `get_pr_files`) and then we can work together to refactor this into minimal Ruby with e.g. `select` and `map`.

**Comment by carlocab** [comment]
> > This feels like something that should be on the pr-automerge layer so it doesn't even try to merge rather than try and fail.
> 
> I thought about it. But this excludes 2 cases:
> 
> * maintainers running pr-pull locally (I don't know how often that happens)

This only happens once in every few thousand PRs. I don't think it's worth considering this case.

> * automerge-skip label, which will not use the pr-automerge process

This happens reasonably often.

**Comment by Bo98** [comment]
> * automerge-skip label, which will not use the pr-automerge process

Sounds like this would fit better as a triage job that assigns a new label that pr-automerge can check and it makes it obvious why the PR isn't being automerged. Can also make it assign a reviewer to the person that added the conflict-whatever label to the large PR.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb:27]: When the automerge in homebrew-core fails due to the conflict check: you need to manually pr-pull with that flag to overcome the automerge failure. Of course you need to review the conflict manually:
- is the corresponding long build PR currently running / being worked on
- is the corresponding long build PR stale? If yes, maybe add a comment there that you merged something and a rebase will be needed.

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/pr-pull.rb:27]: It feels like a nicer solution here might be to remove the label instead?

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb]: Done

File: Library/Homebrew/utils/github.rb
Comment (iMichka) [reply Library/Homebrew/utils/github.rb]: I moved this out of the GitHub file

File: Library/Homebrew/utils/github.rb
Comment (iMichka) [reply Library/Homebrew/utils/github.rb]: I moved this out of the GitHub file

File: Library/Homebrew/utils/github.rb
Comment (iMichka) [reply Library/Homebrew/utils/github.rb]: Done

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb:403]: I get this:
```
Error: You are trying to merge a pull request that conflicts with a long running build in
https://github.com/Homebrew/homebrew-core/pull/9000
https://github.com/Homebrew/homebrew-core/pull/105565
```

If I add the extra space in front, I get this, which looks worse:
```
Error: You are trying to merge a pull request that conflicts with a long running build in
  https://github.com/Homebrew/homebrew-core/pull/9000
https://github.com/Homebrew/homebrew-core/pull/105565
```

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/pr-pull.rb:403]:
```suggestion
        #{conflicts * "  \n. "}
```
I think this is what you want

File: Library/Homebrew/utils/github.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb:482]:
```suggestion
      get_pull_request_changed_files(name, tap_remote_repo, pr["number"]).any? { |f| f["filename"] == file }
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:389]: `long_build_pr_files` should be a `Hash` rather than an `Array` with the filenames as keys so you can do a faster, key-based lookup rather than having to iterate through it.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [reply Library/Homebrew/dev-cmd/pr-pull.rb:403]: Oh, sorry, it should actually be
```suggestion
        #{conflicts * "\n  "}
```
Spaces are hard

File: Library/Homebrew/utils/github.rb
Comment (carlocab) [reply Library/Homebrew/utils/github.rb]: By "query the files twice" I mean making a REST API call to ask GitHub about the files modified in a pull request.

I actually suspect this will get rate-limited pretty quickly if we make one (and possibly more) API call per PR, since we get only a few thousand of these API calls in a given time period, and there are over a hundred open PRs at any given time.

File: Library/Homebrew/utils/github.rb
Comment (iMichka) [reply Library/Homebrew/utils/github.rb]: Done

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb]: Done

**Comment by iMichka** [comment]
Remaining issues:
- Using a different label than "long build"; what should it be?
- Move part / the whole code as a triage job (@Bo98) ?
- @carlocab having concerns about rate limits of the GithubApi: I think the number of calls done is pretty low, especially if we move to a new label that is not "long build" to pick only a limited set of PRs to check.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:380]: I think you can do something like this

```suggestion
    long_build_pr_files = Hash.new { |h, k| h[k] = [] }
    long_prs.each do |long_pr|
      GitHub.get_pull_request_changed_files(name, tap_remote_repo, long_pr).each do |pr_file|
        long_build_pr_files[pr_file["filename"]] << long_pr
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:403]: Should we show the name of the conflicting files?

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/pr-pull.rb:27]: Can we ship without this flag for now and see if we need it? My concern is that it's so destructive if someone passes it and not visible to anyone else.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:380]:
```suggestion
    long_build_pr_files = search_issues(
      "org:#{name}", repo: tap_remote_repo, state: "open", label: '"long build'",
    ).each_with_object({}) do |pr, hash|
      number = pr["number"]
      GitHub.get_pull_request_changed_files(name, tap_remote_repo, number).each do |file|
        key = file["filename"]
        hash[key] ||= []
        hash[key] << number
```

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/pr-pull.rb:380]: Suggested a more comprehensive refactoring above.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:393]:
```suggestion
    conflicts = this_pr_files.flat_map do |file|
      next unless long_build_pr_files.key?(file)

      long_build_pr_files[file].map do |pr_number|
        "#{tap.default_remote}/pull/#{pr_number}"
      end
    end.compact
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:401]:
```suggestion
      You are trying to merge a pull request that conflicts with a long running build in:
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/pr-pull.rb:403]: Agreed.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb]: Done. I removed the flag.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb]: Done

**Comment by iMichka** [comment]
I had to refactor the code again to add the filenames as asked.

I changed the label to `no long build conflict`.
@carlocab about the rate limit. If we have lets say 3 pull request with that label, it means that each `brew pr-pull` will call the API: once to get the pull requests, and 3 times to get the files list. This means 4 API calls. If we merge 30 pull requests at once (which is already a lot, even for automerge), this means 4*30=120 API calls in a few minutes. Is this too much?

**Comment by iMichka** [comment]
I'm ready for a new round of review. I would like to ship this. If we get issues with the rate limit, we will revert this PR.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:372]:
```suggestion
  def pr_check_conflicts(name, tap_remote_repo, pr)
    hash_template = Proc.new { |h, k| h[k] = [] }
    long_build_pr_files = GitHub.search_issues(
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:378]:
```suggestion
    ).each_with_object(Hash.new hash_template) do |long_build_pr, hash|
      number = long_build_pr["number"]
      GitHub.get_pull_request_changed_files(name, tap_remote_repo, number).each do |file|
        key = file["filename"]
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:391]:
```suggestion
    conflicts = this_pr_files.each_with_object(Hash.new hash_template) do |file, hash|
      filename = file["filename"]
      next unless long_build_pr_files.key?(filename)

      long_build_pr_files[filename].each do |pr_number|
        key = "#{tap_remote_repo}/pull/#{pr_number}"
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-pull.rb:407]:
```suggestion
        #{JSON.pretty_generate(conflicts)}
```

**Review by carlocab — APPROVED** [review]
I'm ok to try this, but I'm still wondering if we should investigate alternative approaches raised in this PR.

Optional suggestions, but it only works if you use them all.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:380]:
```suggestion
    long_build_pr_files = GitHub.search_issues(
      "org:#{name}", repo: tap_remote_repo, state: "open", label: 
      '"no long build conflict"'
    ).each_with_object({}) do |long_build_pr, hash|
      number = long_build_pr["number"]
      GitHub.get_pull_request_changed_files(name, tap_remote_repo, number).each do |file|
        key = file["filename"]
        hash[key] ||= []
        hash[key] << number
      end
```

I think this is a little more readable and is no more lines of code.

**Review by MikeMcQuaid — APPROVED** [review]
Very minor style suggestions but fine for this to 🚢 as-is. If you apply style suggestions as-is: no need for another review.

Nice work @iMichka!

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (carlocab) [reply Library/Homebrew/dev-cmd/pr-pull.rb:380]: Oops, this reverses my earlier suggestion 😅 @iMichka can decide whichever is preferred.

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/pr-pull.rb:380]: Yeh, I don't mind too much either way, just think this way is a little more intuitive to non-Rubyists.

**Comment by dawidd6** [comment]
Publish job failed after those changes:
https://github.com/Homebrew/homebrew-core/runs/7632246079?check_suite_focus=true

Not sure what's wrong here.

**Comment by carlocab** [comment]
#13632

**Comment by p-linnane** [comment]
The changes to `github.rb` cause issues when checking for duplicate pull requests.

Before these changes:

```
brew bump-cask-pr --version=1.2.4 zprint

Error: These pull requests may be duplicates:
Update zprint from 1.2.3 to 1.2.4 https://github.com/Homebrew/homebrew-cask/pull/129070
Duplicate PRs should not be opened. Use --force to override this error.
```

After these changes:

```
brew bump-cask-pr --version=1.2.4 zprint

Error: Not Found
```

**Comment by carlocab** [comment]
Thanks, @p-linnane. Please open an issue and we can look at it.

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (Bo98) [inline Library/Homebrew/dev-cmd/pr-pull.rb:371]:
```suggestion
  def pr_check_conflicts(user, tap_remote_repo, pr)
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (Bo98) [inline Library/Homebrew/dev-cmd/pr-pull.rb:374]:
```suggestion
      "org:#{user}", repo: tap_remote_repo, state: "open", label: "\"no long build conflict\""
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (Bo98) [inline Library/Homebrew/dev-cmd/pr-pull.rb:377]:
```suggestion
      GitHub.get_pull_request_changed_files(tap_remote_repo, number).each do |file|
```

File: Library/Homebrew/dev-cmd/pr-pull.rb
Comment (Bo98) [inline Library/Homebrew/dev-cmd/pr-pull.rb:383]:
```suggestion
    this_pr_files = GitHub.get_pull_request_changed_files(tap_remote_repo, pr)
```

File: Library/Homebrew/utils/github.rb
Comment (Bo98) [inline Library/Homebrew/utils/github.rb:483]:
```suggestion
        tap_remote_repo, pr["number"]
```

File: Library/Homebrew/utils/github.rb
Comment (Bo98) [inline Library/Homebrew/utils/github.rb:506]:
```suggestion
  def get_pull_request_changed_files(tap_remote_repo, pr)
    API.open_rest(url_to("repos", tap_remote_repo, "pulls", pr, "files"))
```

**Pull request merged**
